File: tools4a-toast/src/main/java/com/github/panpf/tools4a/toast/Toastx.java
/*
 * Copyright (C) 2020 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.github.panpf.tools4a.toast;

import android.content.Context;
import android.view.View;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;
import androidx.fragment.app.Fragment;

import com.github.panpf.tools4a.run.Runx;

public class Toastx {

    private Toastx() {
    }

    /* ************************************* Context ***************************************** */


    public static void showLong(@NonNull Context context, @NonNull final CharSequence message) {
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }

    public static void showLong(@NonNull Context context, @StringRes final int messageResId) {
        final CharSequence message = context.getResources().getText(messageResId);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }

    public static void showLongWithFormat(@NonNull Context context, @NonNull final String format, @NonNull final Object... params) {
        final CharSequence message = String.format(format, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }

    public static void showLongWithFormat(@NonNull Context context, @StringRes final int messageResId, @NonNull final Object... params) {
        final CharSequence message = context.getString(messageResId, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }


    public static void showShort(@NonNull Context context, @NonNull final CharSequence message) {
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }

    public static void showShort(@NonNull Context context, @StringRes final int messageResId) {
        final CharSequence message = context.getResources().getText(messageResId);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }

    public static void showShortWithFormat(@NonNull Context context, @NonNull final String format, @NonNull final Object... params) {
        final String message = String.format(format, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }

    public static void showShortWithFormat(@NonNull Context context, @StringRes final int messageResId, @NonNull final Object... params) {
        final CharSequence message = context.getResources().getString(messageResId, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }


    /* ************************************ Fragment **************************************** */


    public static void showLong(@NonNull Fragment fragment, @NonNull final CharSequence message) {
        final Context context = fragment.requireContext();
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }

    public static void showLong(@NonNull Fragment fragment, @StringRes final int messageResId) {
        final Context context = fragment.requireContext();
        final CharSequence message = context.getResources().getText(messageResId);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }

    public static void showLongWithFormat(@NonNull Fragment fragment, @NonNull final String format, @NonNull final Object... params) {
        final Context context = fragment.requireContext();
        final CharSequence message = String.format(format, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }

    public static void showLongWithFormat(@NonNull Fragment fragment, @StringRes final int messageResId, @NonNull final Object... params) {
        final Context context = fragment.requireContext();
        final CharSequence message = context.getResources().getString(messageResId, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }


    public static void showShort(@NonNull Fragment fragment, @NonNull final CharSequence message) {
        final Context context = fragment.requireContext();
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }

    public static void showShort(@NonNull Fragment fragment, @StringRes final int messageResId) {
        final Context context = fragment.requireContext();
        final CharSequence message = context.getResources().getText(messageResId);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }

    public static void showShortWithFormat(@NonNull Fragment fragment, @NonNull final String format, @NonNull final Object... params) {
        final Context context = fragment.requireContext();
        final CharSequence message = String.format(format, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }

    public static void showShortWithFormat(@NonNull Fragment fragment, @StringRes final int messageResId, @NonNull final Object... params) {
        final Context context = fragment.requireContext();
        final CharSequence message = context.getResources().getString(messageResId, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }


    /* ************************************* View ***************************************** */


    public static void showLong(@NonNull View view, @NonNull final CharSequence message) {
        final Context context = view.getContext();
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }

    public static void showLong(@NonNull View view, @StringRes final int messageResId) {
        final Context context = view.getContext();
        final CharSequence message = context.getResources().getText(messageResId);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }

    public static void showLongWithFormat(@NonNull View view, @NonNull final String format, @NonNull final Object... params) {
        final Context context = view.getContext();
        final CharSequence message = String.format(format, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }

    public static void showLongWithFormat(@NonNull View view, @StringRes final int messageResId, @NonNull final Object... params) {
        final Context context = view.getContext();
        final CharSequence message = context.getResources().getString(messageResId, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_LONG).show());
    }


    public static void showShort(@NonNull View view, @NonNull final CharSequence message) {
        final Context context = view.getContext();
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }

    public static void showShort(@NonNull View view, @StringRes final int messageResId) {
        final Context context = view.getContext();
        final CharSequence message = context.getResources().getText(messageResId);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }

    public static void showShortWithFormat(@NonNull View view, @NonNull final String format, @NonNull final Object... params) {
        final Context context = view.getContext();
        final CharSequence message = String.format(format, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }

    public static void showShortWithFormat(@NonNull View view, @StringRes final int messageResId, @NonNull final Object... params) {
        final Context context = view.getContext();
        final CharSequence message = context.getResources().getString(messageResId, params);
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> Toast.makeText(appContext, message, Toast.LENGTH_SHORT).show());
    }


    /* ************************************* Show View ***************************************** */

    /**
     * @deprecated Starting from Android 11, the app cannot send custom toasts in the background
     */
    @Deprecated
    public static void showLongWithView(@NonNull final View view) {
        final Context context = view.getContext();
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> {
            Toast toast = new Toast(appContext);
            toast.setView(view);
            toast.setDuration(Toast.LENGTH_LONG);
            toast.show();
        });
    }

    /**
     * @deprecated Starting from Android 11, the app cannot send custom toasts in the background
     */
    @Deprecated
    public static void showShortWithView(@NonNull final View view) {
        final Context context = view.getContext();
        final Context appContext = context.getApplicationContext();
        Runx.runOnMainThread(() -> {
            Toast toast = new Toast(appContext);
            toast.setView(view);
            toast.setDuration(Toast.LENGTH_SHORT);
            toast.show();
        });
    }
}

File: tools4a-toast-ktx/src/main/java/com/github/panpf/tools4a/toast/ktx/Toastx.kt
/*
 * Copyright (C) 2020 panpf <panpfpanpf@outlook.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@file:Suppress("NOTHING_TO_INLINE")

package com.github.panpf.tools4a.toast.ktx

import android.content.Context
import android.view.View
import androidx.annotation.StringRes
import com.github.panpf.tools4a.toast.Toastx

/*
 * Toast related extension method
 */


/* ************************************* Context ***************************************** */


inline fun Context.showLongToast(message: CharSequence) = Toastx.showLong(this, message)

inline fun Context.showLongToast(@StringRes messageResId: Int) = Toastx.showLong(this, messageResId)

inline fun Context.showLongToastWithFormat(message: String, vararg params: Any) = Toastx.showLongWithFormat(this, message, *params)

inline fun Context.showLongToastWithFormat(@StringRes messageResId: Int, vararg params: Any) = Toastx.showLongWithFormat(this, messageResId, *params)


inline fun Context.showShortToast(message: CharSequence) = Toastx.showShort(this, message)

inline fun Context.showShortToast(@StringRes messageResId: Int) = Toastx.showShort(this, messageResId)

inline fun Context.showShortToastWithFormat(message: String, vararg params: Any) = Toastx.showShortWithFormat(this, message, *params)

inline fun Context.showShortToastWithFormat(@StringRes messageResId: Int, vararg params: Any) = Toastx.showShortWithFormat(this, messageResId, *params)


/* ************************************* androidx.fragment.app.Fragment ***************************************** */


inline fun androidx.fragment.app.Fragment.showLongToast(message: CharSequence) = Toastx.showLong(this, message)

inline fun androidx.fragment.app.Fragment.showLongToast(@StringRes messageResId: Int) = Toastx.showLong(this, messageResId)

inline fun androidx.fragment.app.Fragment.showLongToastWithFormat(message: String, vararg params: Any) = Toastx.showLongWithFormat(this, message, *params)

inline fun androidx.fragment.app.Fragment.showLongToastWithFormat(@StringRes messageResId: Int, vararg params: Any) = Toastx.showLongWithFormat(this, messageResId, *params)


inline fun androidx.fragment.app.Fragment.showShortToast(message: CharSequence) = Toastx.showShort(this, message)

inline fun androidx.fragment.app.Fragment.showShortToast(@StringRes messageResId: Int) = Toastx.showShort(this, messageResId)

inline fun androidx.fragment.app.Fragment.showShortToastWithFormat(message: String, vararg params: Any) = Toastx.showShortWithFormat(this, message, *params)

inline fun androidx.fragment.app.Fragment.showShortToastWithFormat(@StringRes messageResId: Int, vararg params: Any) = Toastx.showShortWithFormat(this, messageResId, *params)


/* ************************************* View ***************************************** */


inline fun View.showLongToast(message: CharSequence) = Toastx.showLong(this, message)

inline fun View.showLongToast(@StringRes messageResId: Int) = Toastx.showLong(this, messageResId)

inline fun View.showLongToastWithFormat(message: String, vararg params: Any) = Toastx.showLongWithFormat(this, message, *params)

inline fun View.showLongToastWithFormat(@StringRes messageResId: Int, vararg params: Any) = Toastx.showLongWithFormat(this, messageResId, *params)


inline fun View.showShortToast(message: CharSequence) = Toastx.showShort(this, message)

inline fun View.showShortToast(@StringRes messageResId: Int) = Toastx.showShort(this, messageResId)

inline fun View.showShortToastWithFormat(message: String, vararg params: Any) = Toastx.showShortWithFormat(this, message, *params)

inline fun View.showShortToastWithFormat(@StringRes messageResId: Int, vararg params: Any) = Toastx.showShortWithFormat(this, messageResId, *params)


/* ************************************* Show View ***************************************** */


inline fun View.showLongToastWithSelf() = Toastx.showLongWithView(this)

inline fun View.showShortToastWithSelf() = Toastx.showShortWithView(this)